repoaudit/repoaudit/utils.py

```diff
@@ -26,14 +41,212 @@ def hexdigest(self, alg: str) -> str:
+
+
+def destroy_gpg(gpg: Optional[gnupg.GPG], keep_folder: bool = False) -> None:
+    """Cleans up the resources used by a GPG object"""
+    if gpg and Path(gpg.gnupghome).exists():
+        if keep_folder:
+            home_path = Path(gpg.gnupghome)
+            for filename in home_path.iterdir():
+                file_path = home_path / filename
+                if file_path.is_file() or file_path.is_symlink():
+                    file_path.unlink()
+                elif file_path.is_dir():
+                    shutil.rmtree(file_path)
+        else:
+            shutil.rmtree(gpg.gnupghome)
+
+
+def check_signature(repo: str, dist: str, file_url: str,
+                    gpg: gnupg.GPG, errors: RepoErrors,
+                    signature_url: Optional[str] = None,
+                    verify: Optional[str] = None) -> bool:
+    """
+    Check the signature on a file. If the signature is detached (in a separate file)
+    its url can be specified with signature_url. It is expected input gpg already has
+    the public keys loaded.
+    """
+
+    try:
+        file_text = get_url(file_url, verify=verify).text
+        if signature_url:
+            sig_text = get_url(signature_url, verify=verify).text
+            sig_file_loc = Path(gpg.gnupghome) / "temp_sig_file.gpg"
+
+            file = open(sig_file_loc, "w")
+            file.write(sig_text)
+            file.close()
+
+            verified = gpg.verify_data(str(sig_file_loc), file_text.encode())
+        else:
+            verified = gpg.verify(file_text)
+
+        if not verified:
+            errors.add(
+                repo, dist,
+                f"Signature verification failed for {file_url} " +
+                (f"with the signature {signature_url}" if signature_url else "")
+            )
+            return False
+    except HTTPError as e:
+        errors.add(
+            repo, dist,
+            f"While checking signatures, could not access file at {e.response.url}: {e}"
+        )
         return False
+
+    return True
+
+
+def output_result(errors: RepoErrors, file_name: Optional[str]) -> bool:
+    """Output number of packages processed and errors."""
+
+    if file_name is not None:
+        text_output = errors.get_json()
+        file = open(file_name, "w")
+        file.write(text_output)
+        file.close()
     else:
-        click.echo("No errors detected.")
+        text_output = errors.get_output()
+        click.echo(text_output)
+    return errors.error_count() == 0
+
+
+def package_output(proc_packages: int) -> None:
+    click.echo(f"Checked {proc_packages} package(s).")
+
+
+def check_repo_empty(url: str, verify: Optional[str] = None) -> bool:
+    """Returns true if repo is empty, false otherwise."""
+    try:
+        resp = get_url(url, verify=verify)
+        content = re.findall(r"href=[\"'](.*)[\"']", resp.text)
+        content = list(filter(lambda c: ".." not in c, content))
+        return len(content) == 0
+    except HTTPError:
```

Comment:
This would check for empty repositories such as the following.
https://packages.microsoft.com/repos/cbl-o/
https://packages.microsoft.com/repos/UPS-demo/
https://packages.microsoft.com/yumrepos/amlfs-el8/

Then they can be reported as empty.